Sources/Testing/Parameterization/CustomTestArgument.swift

```diff
@@ -0,0 +1,73 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+//
+
+/// A protocol for customizing how arguments passed to parameterized tests are
+/// represented.
+@_spi(ExperimentalParameterizedTesting)
+public protocol CustomTestArgument: Sendable {
+  /// Get the ID of this test argument, using the provided context.
+  ///
+  /// - Parameters:
+  ///   - context: Context about this argument which may be useful in forming
+  ///     its ID.
+  ///
+  /// - Returns: The ID of this test argument scoped to the provided context
+  ///
+  /// The ID of a test argument should be stable and unique in order to allow
+  /// re-running specific test cases of a parameterized test function. It does
+  /// not need to be human-readable.
+  ///
+  /// By default, the testing library derives an ID for a test argument using
+  /// `String(describing:)`. The resulting string may not be stable or unique.
+  /// If the type of the argument conforms to `Identifiable` and its associated
+  /// `ID` type is `String`, the value of calling its `id` property is used
+  /// instead.
+  ///
+  /// It is possible that neither of the approaches for obtaining an ID detailed
+  /// above provide sufficient stability or uniqueness. If the type of the
+  /// argument is made to conform to ``CustomTestArgument``, then the ID
+  /// returned by this method is used.
+  func argumentID(in context: Test.Case.Argument.Context) -> String
```

Comment:
We have several `Context` types already. Seems like a rethink might be appropriate here.

Reply:
The concept of "a context" to provide additional info is pretty broad, and these types we have serve different purposes. I nested this one to keep it fairly out-of-the-way, so I'm not too concerned by having multiple of these for different roles.